### PR TITLE
links information according to batman protocol

### PR DESCRIPTION
**Pull request by javierbrk**
This module provides per interface links information associated with batman protocol basically using batclt nj command as source. The output is similar to the wifi_links_info to make data parsing easy.

This module depends on batctl-default witch is in use by kmod-batman-adv and  batman-adv-auto-gw-mode. 

This is a sample output of the command  `shared-state-async get bat_links_info` run in one of two librerouters.
```JSON
{
    "lrprimero": [
        {
            "dst_mac": "02:29:0f:1f:73:aa",
            "last_seen_msecs": 770,
            "tq": 238,
            "src_mac": "02:29:0f:1f:72:3a",
            "iface": "eth1-2_250"
        },
        {
            "dst_mac": "02:db:d6:fc:3a:bd",
            "last_seen_msecs": 510,
            "tq": 155,
            "src_mac": "02:29:0f:1f:72:3a",
            "iface": "eth1-2_250"
        }
    ],
    "lrsegundo": [
        {
            "dst_mac": "02:db:d6:fc:3a:bd",
            "last_seen_msecs": 480,
            "tq": 255,
            "src_mac": "02:29:0f:1f:73:aa",
            "iface": "eth1-2_250"
        },
        {
            "dst_mac": "02:29:0f:1f:72:3a",
            "last_seen_msecs": 750,
            "tq": 254,
            "src_mac": "02:29:0f:1f:73:aa",
            "iface": "eth1-2_250"
        }
    ]
}
```

### CODE REVIEW

**Comment by ilario** [comment]
Great!

Just one comment:

$RANDOM will not be available in the upcoming LibreMesh releases.
See https://github.com/libremesh/lime-packages/issues/800 and https://github.com/libremesh/lime-packages/pull/991 (this means that the makefile should include a dependency from the replacement package and the uci-default package should include the random-numgen command instead of the $RANDOM)

File: packages/shared-state-bat_links_info/tests/test_shared-state-bat_links_info.lua [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    it('a simple test to get node info and assert requiered fields are present', function()
+        local links_info = {}
+        links_info = get_bat_links_info()
+        assert.are.equal(table.concat(iwinfo.mocks.wlan1_mesh_mac,":"), links_info[1].src_mac)
+        assert.are.equal('02:58:47:da:4e:aa', links_info[1].dst_mac)
+        assert.are.equal(26, links_info[1].hard_ifindex)
+        assert.are.equal(1990, links_info[1].last_seen_msecs)
+        assert.are.equal("wlan0-mesh_250", links_info[1].iface)
+    end)
+end)
```
Comment (G10h4ck) [inline packages/shared-state-bat_links_info/tests/test_shared-state-bat_links_info.lua:27]: New line missing at the end of file

**Review by G10h4ck — CHANGES_REQUESTED** [review]
`hard_ifindex`should be removed, it is an internal index not useful for other nodes or the used, and better to not spread it around. Plus see my comments inline.

File: packages/shared-state-bat_links_info/tests/test_shared-state-bat_links_info.lua (hunk 1 above)
Comment (javierbrk) [reply packages/shared-state-bat_links_info/tests/test_shared-state-bat_links_info.lua]: done !

**Comment by javierbrk** [comment]
The last push includes the tq parameter for every neighbor

File: packages/shared-state-bat_links_info/files/usr/bin/shared-state-publish_bat_links_info [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	for key,neight_value in pairs (batneights_obj) do
+		local macparts = network.get_mac(neight_value.iface)
+		local src_macaddr = table.concat(macparts,":")
+		neight_value.hard_ifindex=nil
+		neight_value.src_mac=src_macaddr
+		for key,link_value in pairs (batlinks_obj) do
+			if link_value.hard_ifname == neight_value.iface and 
+			link_value.neigh_address== link_value.orig_address and  
+			link_value.neigh_address== neight_value.dst_mac then
+				neight_value.tq = link_value.tq
```
Comment (G10h4ck) [inline packages/shared-state-bat_links_info/files/usr/bin/shared-state-publish_bat_links_info:48]: Let's add a comment explaining that `tq` stands for Batman Transmit Quality

**Review by G10h4ck — CHANGES_REQUESTED** [review]
See inline comments

File: packages/shared-state-bat_links_info/files/usr/bin/shared-state-publish_bat_links_info [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+local JSON = require("luci.jsonc")
+local utils = require('lime.utils')
+local network = require ("lime.network")
+
+local hostname = utils.hostname()
+
+function get_bat_links_info()
+	local batneights_obj={}
+	local batlinks_obj={}
+	local batlinks = utils.unsafe_shell("batctl oj")
```
[the remark this answers is not in the document]
Comment (javierbrk) [reply packages/shared-state-bat_links_info/files/usr/bin/shared-state-publish_bat_links_info]: done

File: packages/shared-state-bat_links_info/files/usr/bin/shared-state-publish_bat_links_info (hunk 2 above)
Comment (javierbrk) [reply packages/shared-state-bat_links_info/files/usr/bin/shared-state-publish_bat_links_info]: sure !

**Comment by G10h4ck** [comment]
should be ported to shared-state-async before merging

**Comment by ilario** [comment]
At some point in the future, when we will have BATMAN_V working (the
possibility to select it was introduced in
https://github.com/libremesh/lime-packages/pull/1004 as @lindnermarek said
it is better than BATMAN_IV. But the first tests failed, as reported on
https://github.com/libremesh/lime-packages/issues/1059 . The progress on
this is being tracked here:
https://github.com/libremesh/lime-packages/issues/1009 ), we should check
if this works also for BATMAN_V.

-- 
Inviato dal mio smartstone

**Review by G10h4ck — APPROVED** [review]

**Pull request merged**